### PR TITLE
[CPDEL-713] CIP change alert

### PR DESCRIPTION
**Pull request by FriedSock**
## Ticket and context

Ticket: https://dfedigital.atlassian.net/browse/CPDEL-713

### Is there anything weird that code reviewer should know?
Nope, everything is straightforward.

### Review Checks
- [x] All commit messages are meaningful and true
- [x] Added enough automated tests
- [x] All forms have an `id` attribute on them

### How can someone see it it review app?
1. Click the link to review app (posted by a `github-actions` bot below)
2. Follow the steps from the ticket.

### CODE REVIEW

File: app/controllers/cip_change_alert_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class CipChangeAlertController < ApplicationController
+  before_action :authenticate_user!
+
+  def start
+    @original_cip_name = cip_change_message.original_cip.name
+    @new_cip_name = cip_change_message.original_cip.name
```
Comment (xswoodsx) [inline app/controllers/cip_change_alert_controller.rb:8]: should this be `cip_change_message.new_name.name`??

Comment (FriedSock) [reply app/controllers/cip_change_alert_controller.rb:8]: yes.. I originally had this bug but I fixed it when I changed these into helper methods below. So they don't need to be here any more 😅

File: app/controllers/cip_change_alert_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class CipChangeAlertController < ApplicationController
+  before_action :authenticate_user!
+
+  def start
+  end
+
+  def acknowledge_start
+    current_user.cip_change_message&.destroy!
```
Comment (kicferk1) [inline app/controllers/cip_change_alert_controller.rb:10]: Do we need `current_user` here? `cip_change_message` is a helper method already :)

File: app/forms/cip_change_alert_form.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class CipChangeAlertForm
+  include ActiveModel::Model
+
+  attr_accessor :view_guidance_option
+
+  validates :view_guidance_option, presence: { message: "TODO fill me in, I am not real copy" }
```
Comment (kicferk1) [inline app/forms/cip_change_alert_form.rb:8]: I am not a real copy, but maybe the fact that this needs to be filled in can be stored in JIRA, and we can put a message here that looks reasonable to the user :D 

"Select whether you want to view guidance"?

File: app/views/cip_change_alert/guidance.html.erb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      <li>
+        bullet item
+      </li>
+    </ul>
+
+    <%= form_for :cip_change_alert, url: cip_change_alert_guidance_path, id: "cip-change-alert-form", method: :post do |form| %>
+      <%= form.govuk_submit "Continue" %>
+    <% end %>
+  </div>
+</div>
```
Comment (kicferk1) [inline app/views/cip_change_alert/guidance.html.erb:31]: I love that 😂 

Can we maybe not push that to prod until we get the right copy? :D

File: app/views/cip_change_alert/guidance.html.erb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    <ul class="govuk-list govuk-list--bullet">
+      <li>
+        bullet item
+      </li>
+      <li>
+        bullet item
+      </li>
+    </ul>
+
+    <%= form_for :cip_change_alert, url: cip_change_alert_guidance_path, id: "cip-change-alert-form", method: :post do |form| %>
```
Comment (kicferk1) [inline app/views/cip_change_alert/guidance.html.erb:27]: Let's use `form_with`, https://stackoverflow.com/questions/43868976/rails-5-form-for-vs-form-with

File: spec/cypress/integration/CipChangeAlert.feature [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    Given I am logged in as existing user with email "cip-change-early-career-teacher@example.com"
+    And I am on "dashboard" page
+    When I click on "link" containing "brief guide"
+    Then I should be on "cip change alert guidance" page
+    And I click on "continue button"
+    Then I should be on "training and support" page
+
+    When I navigate to "dashboard" page
+    And I click on "link" containing "brief guide"
+    Then I should be on "training and support" page
```
Comment (kicferk1) [inline spec/cypress/integration/CipChangeAlert.feature:28]: I think each page here should have 'and the page is accessible' and 'and percy is sent a snapshot called xxx' calls. 

Or at least each new page.

File: app/services/register_and_partner_api/sync_users.rb [hunk 7]
```diff
@@ -105,6 +107,24 @@ def self.assign_user_attributes(attributes, user, profile)
       profile.cohort = get_cohort(attributes)
     end
 
+    def self.create_or_destroy_cip_change_message(user, attributes)
```
Comment (kicferk1) [inline app/services/register_and_partner_api/sync_users.rb:110]: User CIPs can be nil - a quirk of how R&P enter data.

When we change from `nil` to something, we probably don't want to create a guidance message. When we change from something to `nil`, eh, I don't know, probably worth asking Isabelle.

File: app/views/cip_change_alert/guidance.html.erb (hunk 4 above)
Comment (FriedSock) [reply app/views/cip_change_alert/guidance.html.erb:31]: yeah, wasn't planning to merge this until the copy is sorted 😂

File: app/services/register_and_partner_api/sync_users.rb (hunk 7 above)
Comment (FriedSock) [reply app/services/register_and_partner_api/sync_users.rb:110]: Thanks, didn't know about this 👍

File: app/helpers/dashboard_helper.rb [hunk 8]
```diff
@@ -8,4 +8,20 @@ def start_button_text(user)
...
     end
   end
+
+  def start_button_path(user)
+    if user.cip_change_message.present?
+      user.participant_profile.guidance_seen? ? cip_change_alert_start_path : cip_path(user.core_induction_programme)
+    else
+      user.participant_profile.guidance_seen? ? cip_path(user.core_induction_programme) : guidance_question_path
+    end
+  end
```
Comment (kicferk1) [inline app/helpers/dashboard_helper.rb:18]: If user didn't see the guidance, we will never make the new `cip_change_message` for them, right?

So if the message is present, we know they saw the guidance, and therefore should go to `cip_change_alert_start_path` without checking anything.

Am I missing anything?

File: app/views/cip_change_alert/start.html.erb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      Your school has changed your training and support programme from <%= original_cip_name %> to <%= new_cip_name %>
+    </p>
+
+    <p>
+      You can no longer access the materials you were previously using
+    </p>
+
+    <p>
+      Speak to your mentor for more information
+    </p>
```
Comment (kicferk1) [inline app/views/cip_change_alert/start.html.erb:16]: Do we want dots at the end of sentences here?

File: app/views/cip_change_alert/start.html.erb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+    <%= form_with model: @cip_change_message, url: cip_change_alert_start_path, id: "cip-change-alert-form", method: :post do |form| %>
+      <%= form.govuk_submit "Continue" %>
+    <% end %>
+  </div>
+</div>
+
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply app/views/cip_change_alert/start.html.erb:23]: Ah, they point at a different action so we can redirect them to a different location after acknowledging the message. Alright, makes sense to have them both then.

File: app/views/cip_change_alert/start.html.erb (hunk 9 above)
Comment (xswoodsx) [reply app/views/cip_change_alert/start.html.erb:16]: I thought that but it's not on the design.

Comment (xswoodsx) [reply app/views/cip_change_alert/start.html.erb:16]: tell a lie, yes it is. My bad

File: app/helpers/dashboard_helper.rb (hunk 8 above)
Comment (xswoodsx) [reply app/helpers/dashboard_helper.rb:18]: ok, yeah that's right, i hadn't read the user sync properly enough.

Comment (xswoodsx) [reply app/helpers/dashboard_helper.rb:18]: i'll switch it back.

**Review by kicferk1 — APPROVED** [review]

**Pull request merged**